contrib/checkstyle/src/mill/contrib/checkstyle/CheckstyleModule.scala

```diff
@@ -0,0 +1,183 @@
+package mill
+package contrib.checkstyle
+
+import mill.api.{Loose, PathRef}
+import mill.scalalib.{DepSyntax, JavaModule, ScalaModule}
+import mill.util.Jvm
+
+import javax.xml.transform.TransformerFactory
+import javax.xml.transform.stream.{StreamResult, StreamSource}
+
+trait CheckstyleModule extends JavaModule {
+
+  import CheckstyleModule._
+
+  def checkstyle: T[Seq[PathRef]] = T {
+
+    val dst = T.dest
+
+    val classpath = checkstyleClasspath().map(_.path)
+    val options = checkstyleOptions()
+    val config = checkstyleConfig()
+    val format = checkstyleFormat()
+    val report = dst / checkstyleReport()
+    val inputs = sources().map(_.path.toString())
+    val _throw = checkstyleThrow()
+    val transforms = checkstyleTransforms()
+    val transformer = checkstyleTransformer(format)
+
+    val args = options ++
+      Seq("-c", config.path.toString(), "-f", format, "-o", report.toString()) ++
+      inputs
+
+    T.log.info(s"generating checkstyle $format report ...")
+    T.log.debug(s"running checkstyle with $args")
+
+    val errs = Jvm.callSubprocessUnchecked(
+      mainClass = "com.puppycrawl.tools.checkstyle.Main",
+      classPath = classpath,
+      mainArgs = args,
+      workingDir = T.dest
+    ).exitCode
+
+    if (errs == 0) {
+      T.log.info(s"checkstyle found no error, details in $report")
+    } else if (errs > 0 && os.exists(report)) {
+      val msg = s"checkstyle found $errs error(s), details in $report"
+      if (_throw) {
+        throw new RuntimeException(msg)
+      } else {
+        T.log.error(msg)
+      }
+    } else {
+      throw new RuntimeException(s"checkstyle exit($errs)")
+    }
+
+    val reports = transformer.fold(Seq.empty[PathRef]) { transformer =>
+      val f = transformer(report)
+      transforms.map {
+        case (ref, rel) =>
+          val transform = ref.path
+          val out = dst / rel
+          T.log.info(s"transforming checkstyle report using $transform")
+          f(transform, out)
+          T.log.info(s"transformed checkstyle report to $out")
+          PathRef(out)
+      }.toSeq
+    }
+
+    PathRef(report) +: reports
+  }
+
+  def checkstyleClasspath: T[Loose.Agg[PathRef]] = T {
+    defaultResolver().resolveDeps(
+      Agg(ivy"com.puppycrawl.tools:checkstyle:${checkstyleVersion()}")
+    )
+  }
+
+  def checkstyleConfig: T[PathRef] = T.source {
+    checkstyleDir().path / "config.xml"
+  }
+
+  def checkstyleDir: T[PathRef] = T {
+    PathRef(millSourcePath / "checkstyle")
+  }
```

Comment:
This shouldn't be a `PathRef`, since the downstream `T.source`s will wrap things in `PathRef` for us

Reply:
I think this is not applicable now that i've gone through your code again